src/createEventEmitterProxy.ts

```diff
@@ -0,0 +1,116 @@
+// This package relies on the EventEmitter type.
+// eslint-disable-next-line import/no-nodejs-modules
+import type { EventEmitter } from 'events';
+
+import type { SwappableProxy } from './types';
+
+const filterNoop = () => true;
+const internalEvents: (string | symbol)[] = ['newListener', 'removeListener'];
+const externalEventFilter = (name: string | symbol) =>
+  !internalEvents.includes(name);
+
+/**
+ * Creates a proxy object that initially points to the given object but whose
+ * target can be substituted with another object later using its `setTarget`
+ * method. In addition, when the target is changed, event listeners which have
+ * been attached to the target will be detached and migrated to the new target.
+ *
+ * @param initialTarget - The initial object you want to wrap.
+ * @param opts - The options.
+ * @param opts.eventFilter - Usually, listeners for all events will be migrated
+ * from old targets to new targets, but this function can be used to select
+ * which events you want to migrate. If you pass `'skipInternal'`, then
+ * `newListener` and `removeListener` will be excluded.
+ * @returns The proxy object.
+ */
+export function createEventEmitterProxy<T extends EventEmitter>(
+  initialTarget: T,
+  {
+    eventFilter: givenEventFilter = filterNoop,
+  }: {
+    eventFilter?: ((eventName: string | symbol) => boolean) | 'skipInternal';
+  } = {},
+): SwappableProxy<T> {
+  // parse options
+  const eventFilter =
+    givenEventFilter === 'skipInternal'
+      ? externalEventFilter
+      : givenEventFilter;
+
+  if (typeof eventFilter !== 'function') {
+    throw new Error('createEventEmitterProxy - Invalid eventFilter');
+  }
+
+  let target = initialTarget;
+
+  /**
+   * Changes the object that the proxy wraps.
+   *
+   * @param newTarget - The new object.
+   */
+  let setTarget = (newTarget: T) => {
+    const oldTarget = target;
+    target = newTarget;
+    // migrate listeners
+    oldTarget
+      .eventNames()
+      .filter(eventFilter)
+      .forEach((name) => {
+        oldTarget.rawListeners(name).forEach((handler) => {
+          // @ts-expect-error `rawListeners` returns `Function[]`, but `on`
+          // takes `(...args: any[]) => void`.
+          newTarget.on(name, handler);
+        });
+      });
+    // remove old
+    oldTarget.removeAllListeners();
+  };
+
+  const proxy = new Proxy<T>(target, {
+    get(
+      localTarget: T & { [key: string | symbol]: unknown },
+      name: 'setTarget' | keyof T | string | symbol,
```

Comment:
I decided to throw in `string | symbol` here (and have the target extend `{ [key: string | symbol]: unknown }` as per your suggestion). I thought about it some more, and while it's true that the only properties that will ultimately be readable/writable on the proxy object are those of the underlying object and `setTarget`, that protection is already covered by the `SwappableProxy` and `T` types. When we are building the Proxy and specifying the `get` and `set` options to the Proxy constructor, however, we are a level below the surface, so we need to understand what a proxy fundamentally does: it traps access to _every_ property, not just the ones we care about. So while the fact that the type for the Proxy constructor doesn't let you customize the return type, the types for `get` and `set` at least seem fine to me after all.

Reply:
Oh, actually, this won't work, because we can't make use of the `target` argument here...